src/types/custom.rs

```diff
@@ -1,37 +1,46 @@
 //! Opaque types, used to represent a user-defined [`SimpleType`].
+//!
+//! [`SimpleType`]: super::SimpleType
 use smol_str::SmolStr;
 use std::fmt::{self, Display};
 
-use super::{ClassicType, SimpleType, TypeRow};
+use super::{type_param::TypeArg, ClassicType};
 
-/// An opaque type element. Contains an unique identifier and a reference to its definition.
-//
-// TODO: We could replace the `Box` with an `Arc` to reduce memory usage,
-// but it adds atomic ops and a serialization-deserialization roundtrip
-// would still generate copies.
-#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
+/// An opaque type element. Contains the unique identifier of its definition.
+#[derive(Debug, PartialEq, Eq, Clone, serde::Serialize, serde::Deserialize)]
 pub struct CustomType {
     /// Unique identifier of the opaque type.
+    /// Same as the corresponding [`TypeDef`]
+    ///
+    /// [`TypeDef`]: crate::resource::TypeDef
     id: SmolStr,
-    params: Box<TypeRow>,
+    /// Arguments that fit the [`TypeParam`]s declared by the typedef
+    ///
+    /// [`TypeParam`]: super::type_param::TypeParam
+    params: Vec<TypeArg>,
 }
 
 impl CustomType {
     /// Creates a new opaque type.
-    pub fn new(id: SmolStr, params: impl Into<TypeRow>) -> Self {
+    pub fn new(id: impl Into<SmolStr>, params: impl Into<Vec<TypeArg>>) -> Self {
         Self {
-            id,
-            params: Box::new(params.into()),
+            id: id.into(),
+            params: params.into(),
         }
     }
 
+    /// Creates a new opaque type with no parameters
+    pub const fn new_simple(id: SmolStr) -> Self {
+        Self { id, params: vec![] }
+    }
+
     /// Returns the unique identifier of the opaque type.
     pub fn id(&self) -> &str {
         &self.id
     }
 
     /// Returns the parameters of the opaque type.
-    pub fn params(&self) -> &TypeRow {
+    pub fn params(&self) -> &Vec<TypeArg> {
```

Comment:
Or a `Cow<'_, [TypeArg]>` if you want both :)

But as you say in the other comment, this could be a consideration for the public release.

Reply:
Hmmmm, so I've switched this (and an argument type that was `&Vec<TypeArg>` too) to `&[TypeArg]` as that seems the most general thing.

Is there any advantage in declaring the return type `Cow` (as you suggest) rather than the caller constructing their own `Cow` around the `&[TypeArg]` here? The answer may be, "maybe but only if" we construct+store the `Cow` internally and share it between multiple calls (all returning the same `Cow`), rather than many clients each building their own `Cow` around the same `&TypeArg`, but I'm not even sure about that - isn't building a `Cow` zero-cost anyway?